File: leetcode/HashTable/leetcode-149/149.java
import java.util.Map;

/**
 * Definition for a point.
 * class Point {
 *     int x;
 *     int y;
 *     Point() { x = 0; y = 0; }
 *     Point(int a, int b) { x = a; y = b; }
 * }
 */
class Solution {
    public int maxPoints(Point[] points) {
        if (points == null || points.length == 0) {
            return 0;
        }

        int res = 0;
        for (int i = 0; i < points.length; i++) {
            Map<String, Integer> map = new HashMap<>();
            int duplicate = 1;
            for (int j = i + 1; j < points.length; j++) {
                if (points[i].x == points[j].x && points[i].y == points[j].y) {
                    duplicate++;
                    continue;
                }
                int dx = points[i].x - points[j].x;
                int dy = points[i].y - points[j].y;
                int d = gcd(dx, dy);
                String p = (dx / d) + "#" + (dy / d);
                map.put(p, map.getOrDefault(p, 0) + 1);
            }
            res = Math.max(res, duplicate);

            for (String key : map.keySet()) {
                res = Math.max(res, map.get(key) + duplicate);
            }
        }
        
        return res;
    }

    private int gcd(int x, int y) {
        return y == 0 ? x : gcd(y, x % y);
    }
}

File: leetcode/HashTable/leetcode-149/149.go
package hashtable

import "strconv"

// Point : Definition for a point.
type Point struct {
	X int
	Y int
}

func maxPoints(points []Point) int {
	if points == nil || len(points) == 0 {
		return 0
	}

	res := 0
	for i := 0; i < len(points); i++ {
		m := make(map[string]int)
		dup := 1
		for j := i + 1; j < len(points); j++ {
			if points[i].X == points[j].X && points[i].Y == points[j].Y {
				dup++
				continue
			}
			dx, dy := points[i].X-points[j].X, points[i].Y-points[j].Y
			d := gcd(dx, dy)
			p := strconv.Itoa(dx/d) + "#" + strconv.Itoa(dy/d)
			m[p] = m[p] + 1
		}
		res = max(res, dup)
		for _, v := range m {
			res = max(res, v+dup)
		}
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func gcd(x, y int) int {
	if y == 0 {
		return x
	}
	return gcd(y, x%y)
}
